### ISSUE TITLE
Give Lokad/ILPack a test drive to see whether we can use it to improve test coverage

### ISSUE THREAD

**Issue by stakx** [castleproject/Core#431]
Quoting @jonorossi from https://github.com/dotnet/corefx/issues/4491#issue-116708679:

> `AssemblyBuilder.Save` and `AssemblyBuilderAccess.RunAndSave` isn't available in .NET Core [...].
>
> Our use case in Castle DynamicProxy is to write out dynamically created assemblies to disk so we can run peverify over the assembly in unit tests. It also greatly helps writing out the assembly and opening it in ildasm to manually verify IL.
>
> Is there any chance we can get this functionality that .NET Framework has introduced into .NET Core, or an alternative way of doing the same thing. This might not sound like a big deal if you aren't familiar with Castle DynamicProxy's internals, but it really is as it heavily reduces our confidence in both our code and the .NET runtime as well as reducing our ability to track down defects.

And @vermorel's response in https://github.com/dotnet/corefx/issues/4491#issuecomment-469009414:

> We have finally rolled our own open-source version of `AssemblyBuilder.Save` based on `System.Reflection.Metadata` (and no further dependencies). It targets .NET Standard and works under .NET Core. It does support dynamically generated assemblies.

This libray sounds promising. I've been busy with other things for a while and don't have all the details about our build script / test process fresh in my mind, but perhaps we could give this library a test drive to see if we can improve test coverage for the non-`netfx` builds.

**Comment by jonorossi** [comment]
I saw that comment in the corefx issue too and had a very quick look at the ILPack code. It is definitely interesting, probably not something I'd want to rely on in production yet but definitely could be a candidate for unit tests.

The other thing I saw recently is the ILVerify project, a new implementation of an IL verifier written in C# and so it isn't tied to .NET Framework, and it has a CLI and library:
- https://github.com/dotnet/corert/tree/master/src/ILVerify
- https://github.com/dotnet/corert/tree/master/src/ILVerification

**Comment by stakx** [comment]
> probably not something I'd want to rely on in production yet 

Yes, I came to the same conclusion. From the few tests that I ran, I can see two immediate problems:

 1. Lokad.ILPack doesn't support the same target platforms that we do. In fact, it supports only a single target: `netcoreapp2.1`. This seems too restrictive since we're likely aiming for `netstandard2.1` instead.

 2. It produce invalid metadata. (It didn't take any involved experiments to discover this, but a very basic scenario.)

These are roadblocks that may disappear in the future, but for now Lokad.ILPack unfortunately isn't an option. Closing, at least for the time being.
